compiler/rustc_parse/src/parser/item.rs

```diff
@@ -59,10 +59,18 @@ impl<'a> Parser<'a> {
 
         let post_attr_lo = self.token.span;
         let mut items = ThinVec::new();
-        while let Some(item) = self.parse_item(ForceCollect::No)? {
-            self.maybe_consume_incorrect_semicolon(Some(&item));
+
+        // There shouldn't be any semicolons before or afer items.
+        // `parse_item` consumes the appropriate semicolons so any leftover is an error.
+        let mut last = None;
+        while let Some(item) = {
+            self.maybe_consume_incorrect_semicolon(last);
+            self.parse_item(ForceCollect::No)?
+        } {
             items.push(item);
+            last = items.last().map(|x| &**x);
         }
+        self.maybe_consume_incorrect_semicolon(last);
```

Comment:
This may not be needed

Reply:
It's needed to parse trailing semis:
```rs
fn main(){}
;
```

Reply:
The preceding loop should take care of trailing semicolons as well. For this code:
```rust
fn main(){}
;
```
it should parse the item `main` in the first iteration and then iterate again and consume the semicolon.